docs(lane_change): update lane change documentation (#9949)
autowarefoundation/autoware.universe: planning/behavior_path_planner/autoware_behavior_path_lane_change_module/README.md

if (Is dist to next regulatory element < max prepare dist) then (yes)
  #LightPink:False;
  stop
else (no)
end if

#LightGreen:True;
stop

@enduml
```

### Ready Conditions

- Valid lane change path is found.
- Lane change path is safe; does not collide with other dynamic objects.
- Lane change candidate path is approved by an operator.

## Implementation

Lane change module uses a sampling based approach for generating a valid and safe lane changing trajectory. The process for generating the trajectory includes object filtering, metrics sampling, candidate paths generation, and lastly candidate paths evaluation.
Additionally the lane change module is responsible for turn signal activation when appropriate, and inserting a stop point when necessary.

### Global Flowchart

The following diagram, illustrates the overall flow of the lane change module implementation.

```plantuml
@startuml

start

#LightBlue:Update lane change data;
#LightBlue:Filter detected objects;
if (Is lane change required?) then (yes)
  #LightBlue:Activate turn signal;
  #LightBlue:Generate metrics samples;
  if (Is valid metrics available) then (yes)
    While (Can generate new candidate path?) is (TRUE)

>       if (Candidate path is valid & safe?) then (yes)

Reviewer: ```suggestion
      if (Candidate path is valid AND safe?) then (yes)
```

        #LightBlue:Generate drivable area;
        #LightBlue:Execute lane change path;
        while (Can module transit to success state?) is (False)

>           if (Can module trasit to failure state?) then (yes)

Reviewer: ```suggestion
          if (Can module transit to failure state?) then (yes)
```

>             if (Can cancel or abort?) then (yes)

Reviewer: ```suggestion
            if (Can cancel OR abort?) then (yes)
```

The lane-changing phase consists of the shifted path that moves ego from current lane to the target lane. Total duration of lane-changing phase is computed from the `shift_length`, `lateral_jerk` and `lateral_acceleration`.

In principle, positive longitudinal acceleration is considered during lane-changing phase, and is computed as follows.

```C++
lane_changing_acceleration = std::clamp((max_path_velocity - initial_lane_changing_velocity) / lane_changing_time,
  0.0, prepare_longitudinal_acc);
```

Where `max_path_velocity` is the current path speed limit.

!!! warning

    If the longitudinal acceleration of prepare phase is negative (slowing down), AND ego is near terminal, then the lane-changing longitudinal acceleration will also be negative and its value is decided by the parameter `lane_changing_decel_factor`.

The lane changing distance is then computed as follows.

```C++
lane_changing_distance = initial_lane_changing_velocity * lane_changing_duration + 0.5 * lon_acceleration * lane_changing_duration^2
```

The `backward_length_buffer_for_end_of_lane` is added to allow some window for any possible delay, such as control or mechanical delay during brake lag.

> #### Multiple candidate path samples (prepare duration)

Reviewer: Need to put a warning, shorter prepare duration will reduced the "window" for cancel.

Reviewer: Link to this part of the doc https://autowarefoundation.github.io/autoware.universe/main/planning/behavior_path_planner/autoware_behavior_path_lane_change_module/#evaluating-ego-vehicles-position-to-prevent-abrupt-maneuvers

Author: done [8ae3527](https://github.com/autowarefoundation/autoware.universe/pull/9949/commits/8ae35270cd1bccce5ec4f09a3d41a5036feb99bc)

else (covered)
endif

:Calculate total length and goal related distances;
if (total lane change length considering single lane change > distance from current pose to end of current lanes) then (yes)
  #LightPink:Reject path;
  stop
else (no)
endif

if (goal is in current lanes) then (yes)
  if (total lane change length considering multiple lane changes > distance from ego to goal along current lanes) then (yes)
      #LightPink:Reject path;
    stop
  else (no)
  endif
else (no)
endif

if (target lanes is empty) then (yes)
  #LightPink:Reject path;
  stop
else (no)
endif
if (total lane change length considering multiple lane changes  > distance from ego to the end of target lanes) then (yes)
  #LightPink:Reject path;
  stop
else (no)
endif

#LightGreen:Valid Candidate Path;
stop

@enduml
```

!!! warning

    A valid path does NOT mean that the path is safe, however it will be available as a candidate path and can be force approved by operator. A path needs to be both valid AND safe to be automatically approved.

#### Candidate Path Safety

A candidate path is considered safe if:

> 1. There is no overtaking turn lane object (see Overtaking Object Check section below)

Reviewer: The closest explanation is "there's no overtaking object when ego vehicle exited turn direction lane"

> 1. There is no overtaking turn lane object (see Overtaking Object Check section below)

Reviewer: ```suggestion
1. There are no overtaking objects when the ego vehicle exits the turn-direction lane. (see Overtaking Object Check section below)
```

> 2. There is no parked vehicle along the target lane ahead of ego (see Delay Lane Change Check section below)

Reviewer: ```suggestion
2. There is no parked vehicle along the target lane ahead of ego (see [Delay Lane Change Check](#delay-lane-change-check))
```

3. The path does NOT cause ego footprint to exceed the target lane opposite boundary
4. The path passes collision safety check (See [safety check utils explanation](../autoware_behavior_path_planner_common/docs/behavior_path_planner_safety_check.md))

#### Overtaking Object Check

When ego is exiting an intersection on a turning lane, there is a possibility that a rear vehicle will attempt to overtake the ego vehicle. Which can be dangerous if ego is also trying to perform a lane change. Therefore lane change module will adopt a more conservative behavior in such situation.

> If the ego vehicle is currently within an intersection on a truning lane, as shown in the figure below, the generated candidate paths will be marked as unsafe.

Reviewer 2: ```suggestion
If the ego vehicle is currently within an intersection on a turning lane, as shown in the figure below, the generated candidate paths will be marked as unsafe.
```
